src/transformers/trainer_seq2seq.py

```diff
@@ -161,11 +161,17 @@ def prediction_step(
             "synced_gpus": True if is_deepspeed_zero3_enabled() else False,
         }
 
-        model_input_names = self.tokenizer.model_input_names if self.tokenizer is not None else ["input_ids"]
-        generation_inputs = {k: v for k, v in inputs.items() if k in model_input_names}
+        # prepare generation inputs
+        # some encoder-decoder models can have varying encder's and thus
+        # varying model input names
+        if hasattr(self.model, "encoder") and self.model.encoder.main_input_name != self.model.main_input_name:
+            generation_inputs = inputs[self.model.encoder.main_input_name]
+        else:
+            generation_inputs = inputs[self.model.main_input_name]
 
         generated_tokens = self.model.generate(
-            **generation_inputs,
+            generation_inputs,
+            attention_mask=inputs.get("attention_mask", None),
```

Comment:
Can you copy that logic to the newly added `KerasMetricCallback` in keras_callbacks? It has a path with `predict_with_generate` that should use the same logic as here. I think we probably can remove the tokenizer argument/attribute of that callback as a result.